doc/Yang Model Guidelines.md

```diff
@@ -0,0 +1,253 @@
+#   YANG MODEL GUIDELINES
+
+### Rev 1.0 ###
+
+### Revision ###
+
+ | Rev |     Date    |       Author       | Change Description                |
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 1.0 | 22 Aug 2019             |      Praveen Chaudhary      | Initial version                   |
+
+
+
+## This document lists the guidelines, which will be used to write Yang Modules for Sonic. These Yang Modules will be primarily based on ABNF Schema of Sonic. 
+
+
+
+### 1. Each primary section of Config DB (i.e a dictionary in config_db.json) for Example, VLAN, VLAN_MEMBER, INTERFACE  in config DB will be mapped to a container in Yang model.
+
+Example:
+
+#### ABNF
+
+```
+"INTERFACE": {
+        ~~"Ethernet112|2a04:fxxx:40:a709::xxx/126": {
+            "scope": "global",
+            "family": "IPv6"~~
+        }
+ }
+```
+will translate to:
+
+####  Yang
+--
+```
+container INTERFACE {
+~~
+~~
+}
+```
+
+### 2. Each leaf in yang module should have same name as corresponding key in Config DB.
+
+Example:
+Leaf names are same PACKET_ACTION, IP_TYPE and PRIORITY.
+
+#### ABNF
+```
+        "NO-NSW-xxx|xxx": {
+            "PACKET_ACTION": "FORWARD",
+            "IP_TYPE": "IPv6ANY",
+            "PRIORITY": "9xxxx20",
+            .....
+            ....
+        },
+```
+
+####  Yang
+```
+            leaf PACKET_ACTION {
+                type xxxx;
+            }
+            leaf IP_TYPE {
+                type xxxx;
+            }
+            leaf PRIORITY {
+                type uint32 {
+                    range xxxx;
+                }
+            }
+```
+
+
+### 3. Hierarchy of the an objects in yang models will be same as for all the objects at same hierarchy in Redis. If any exception is created then it must be recorded properly.
+For Example: 
+
+"Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE should be at same level in yang model too.
+
+#### ABNF
+```
+"VLAN_INTERFACE": {
+        "Vlan100|2a04:f547:45:6709::1/64": {
+            "scope": "global",
+            "family": "IPv6"
+        }
+}
+"ACL_RULE": {
+        "NO-NSW-PACL-V4|DEFAULT_DENY": {
+            "PACKET_ACTION": "DROP",
+            "IP_TYPE": "IPv4ANY",
+            "PRIORITY": "0"
+        }
+}
+```
+####  Yang
+In Yang, "Family" of VLAN_INTERFACE and "IP_TYPE" of ACL_RULE is at same level. 
+```
+container VLAN_INTERFACE {
+        description "VLAN_INTERFACE part of config_db.json";
+        list xxx {
+    ......
+    ......
+            leaf family {
+                type sonic-head:ip-family;
+            }
+        }
+}
+
+container ACL_RULE {
+        description "ACL_RULE part of config_db.json";
+        list xxx {
+        ......
+    ......
+            leaf IP_TYPE {
+                type sonic-head:ip_type;
+            }
+    }
+}
+``` 
+
+### 4.  if an object is used as a key in Redis, it should be a key in yang model, irrespective of need. Exception must be recorded in Yang Model.
```

Comment:
Where & when do we see a scope for exception?

If none, we could skip the "exception ...".

In future, if there ever would be a need, we can revisit.

Reply:
@praveen-li When we give the interface table data to 'libyang' by skipping ip-prefix key, it will throw syntax as all keys should be provided.

Reply:
Right now, we have 2 scenarios, 1.) ip-prefix as key and 2.) VRF configuration where only interface is key.
I feel, this scenario is tricky to handle. 
One option is: We can make if-name + ip-prefix + vrf name as key, Where while translation to YANG (or with default option in YANG), we can populate non specified field as empty, that way key will be unique.

For Example:
"Ethernet1"|""|"Vrf-red"
"Ethernet2"|""|""
"Ethernet0|"11.11.11.1/24"|"",
"Ethernet0|12.12.12.1/24"|"",
"Ethernet3"|""|"" 
All 2 unique and all valid as per: https://github.com/Azure/SONiC/blob/master/doc/sonic-vrf-hld.md#the-schema-changes.

Reply:
@praveen-li Can you explain "with default option in YANG" a bit. Default value in YANG key is ignored as per RFC 7950. What should be the empty values be when we translate ABNF to YANG ?
One option is to have "default 0.0.0.0/0" in YANG as we have been discussing. But user must provide the default value also in REST API call, otherwise RESTCONF server rejects such request since it always expects all key values to be specified. So, if users sends  "0.0.0.0/0" as ip-prefix (may be in YANG 'description' we should mention that user must send default value for creating empty key), we know that it is default value in YANG and hence empty key item should be created in Redis DB.

Reply:
I go for 0.0.0.0/0 for IP addresses. I wonder what we could for in places, where it is empty string.

Please suggest.

One option could be to have a reserved word indicating empty value. e.g. "None"
Can we confirm a way ?

Reply:
I checked that Libyang allows empty keys.

For Example:
            leaf ip-prefix {
                mandatory true;
                type union {
                        type inet:ip-prefix;
                        type string {
                                pattern "";
                       }
                }
           }

and Data:
			{
	        	"interface": "Ethernet8",
	        	"ip-prefix": "",
				"scope": "global",
				"family": "IPv4"
        	},

is acceptable. 
But since default keys are ignored, we need to find a way to auto-populate this.

Reply:
Can you please elaborate on 'where we should/required-to auto-populate'?

Reply:
@praveen-li We don't have a control if REST API client chooses to skip the key and then REST Server will reject the request before we auto populate it inside management framework.

Reply:
@dutta-partha  -- Please share your suggestions on optional-key handling.